### PR TITLE
Nicer docker compose & co for fast dev environment startup

### PR DESCRIPTION
**Pull request by tswfi**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Fix some problems with the current `docker-compose up` for development.  See below for more
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | git clone && docker-compose up
| Fixed ticket?     | Discussion: https://github.com/PrestaShop/PrestaShop/discussions/31969
| Related PRs       | 
| Sponsor company   | 

* Up mysql to version 8
  * much faster startup for db, supported version
* Fix `sed` parameters in docker_run_git.sh
  * does not create `defines.inc.phpe` anymore
  * this same problem seems to be here also: `.github/actions/setup-prestashop-env/action.yml` didn't touch that though
* Up nodejs to version 16
  * fixes changing the package-json.lock files on startup
  * side note version 18 does not work
* Default to dev mode
  * yes, developers want dev mode
* Default to english language
  * most developers speak English (French is still available easily)
* Update readme for "full reset" of your docker environment
  * Can prevent issues like this: https://github.com/PrestaShop/PrestaShop/issues/30979 https://github.com/PrestaShop/PrestaShop/issues/10934

### CODE REVIEW

File: README.md
Comment (kpodemski) [inline README.md]:
```suggestion
To fully reset your repo and get a fresh start, run (be careful: this removes all your extra files):
```

File: README.md
Comment (kpodemski) [inline README.md]:
```suggestion
# clean everything that is not part of the original repository (node_modules, etc.)
```

**Review by kpodemski — COMMENTED** [review]
Looks good!

I'll ping @SharakPL because I know he's been working on some other improvements.

File: .docker/Dockerfile
Comment (kpodemski) [inline .docker/Dockerfile:20]: I remember some issues with node 14 and back office assets, is everything ok on your side if you changed that?

Comment (0x346e3730) [reply .docker/Dockerfile:20]: +1 for the concern, last time I checked there was a lot of errors with Node 16

File: docker-compose.yml
Comment (0x346e3730) [inline docker-compose.yml:13]: MySQL 8 introduces a new default authentification method (see https://dev.mysql.com/blog-archive/upgrading-to-mysql-8-0-default-authentication-plugin-considerations/).
To avoid authentification issues, you should add `command: --default-authentication-plugin=mysql_native_password` under image (or somewhere else in the yml but at the same indentation level than image)

**Review by 0x346e3730 — DISMISSED** [review]

File: docker-compose.yml
Comment (tswfi) [reply docker-compose.yml:13]: In my tests this works fine with the new defaults. 

And as this is for development I don't see a problem with going with the new defaults.

File: .docker/Dockerfile
Comment (tswfi) [reply .docker/Dockerfile:20]: Yeah, I did double check that SF and BO open correctly with these changes.

Maybe others should try to do `docker compose up` with these changes to get more results?

Comment (tswfi) [reply .docker/Dockerfile:20]: Also just checked and most (if not all) CI tasks already use node16 from actions/setup-node

Comment (kpodemski) [reply .docker/Dockerfile:20]: we had some issues with `make assets` on node 16, I'll check your changes in a moment

**Comment by 0x346e3730** [comment]
Checked out your PR, ran

```sh
rm config/settings.inc.php
docker compose down -v
docker compose build --no-cache
docker-compose up --build --force-recreate
```

and I'm getting the following error

image

**Comment by tswfi** [comment]
> 
> and I'm getting the following error
> 

eh, why it didn't run composer install or anything?

after wait-for-it.sh gets connection to mysql it should first run composer install, npm builds and lots of things before the installer starts..

**Comment by kpodemski** [comment]
I'm using M1 Max, and Docker is so slow... 😠 Anyway, I checked the changes, but for some reason, Docker hangs in this place:

https://capture.dropbox.com/hL4FFXNRdSWneHGj

I don't use Docker regularly, but I'm pretty sure I didn't have this issue before, and it doesn't seem to be related to your changes, so I think we need to wait for somebody else to review the PR 😅

**Comment by tswfi** [comment]
> I'm using M1 Max, and Docker is so slow... 😠 Anyway, I checked the changes, but for some reason, Docker hangs in this place:

yeah, M1 and docker is slow. check the rosetta setting in beta features, it helps a bit
image

**Comment by tswfi** [comment]
> > I'm using M1 Max, and Docker is so slow... 😠 Anyway, I checked the changes, but for some reason, Docker hangs in this place:
> 
> yeah, M1 and docker is slow. check the rosetta setting in beta features, it helps a bit 

or not.. the used apache version segfaults with the rosetta setting :D

**Comment by SharakPL** [comment]
> Looks good!
> 
> I'll ping @SharakPL because I know he's been working on some other improvements.

More improvments here #31527

File: docker-compose.yml
Comment (SharakPL) [inline docker-compose.yml:43]: No need for this change IMO. It will be customizable in .env file after #31527 is merged

Comment (tswfi) [reply docker-compose.yml:43]: lets see which one gets merged first :D

Comment (SharakPL) [reply docker-compose.yml:43]: Matters not which is first, but what will it be in the end. Although english is more common than french and dev mode  should be enabled by default for development so this change is still an improvement.

File: docker-compose.yml
Comment (tswfi) [reply docker-compose.yml:13]: I think a few people have tested this and they didn't have problems with the new mysql8 auth method. 

I think we should embrace the new authentication in develop, if there are problems in the future at least we will know before it becomes a problem in production environments.

File: .docker/Dockerfile
Comment (tswfi) [reply .docker/Dockerfile:20]: Just to resolve this conversation: All CI tasks already use node 16, the dev environment should also use node16. And I haven't seen any reports about problems with make assets in this PR.

**Comment by tswfi** [comment]
@boubkerbribri sorry to ping you but I would have the `mailcatcher` addition PR ready to go as soon as this gets merged :)

**Comment by FabienPapet** [comment]
@tswfi can you please add the empty line at the end of the docker file ? 

Thank you

**Comment by tswfi** [comment]
> @tswfi can you please add the empty line at the end of the docker file ?

added in [b42e40e](https://github.com/PrestaShop/PrestaShop/pull/31976/commits/b42e40e72d11b5fd166b16f8b02befeba97b2225)

**Comment by boubkerbribri** [comment]
> @boubkerbribri sorry to ping you but I would have the `mailcatcher` addition PR ready to go as soon as this gets merged :)

You want to add a mailcatcher for tests ? we already have maildev that starts and close on tests files.

**Comment by boubkerbribri** [comment]
Also : could be nice to have @PrestaShop/qa-automation tests this PR to check that all is good on the nightly ?

**Comment by tswfi** [comment]
> You want to add a mailcatcher for tests ? we already have maildev that starts and close on tests files.

this is not for tests, this is for developers who want to start working easily with minimal fuss by doing basically:

```
git clone ...
docker-compose up
```

and get a working environment where they can test their changes.

**Comment by boubkerbribri** [comment]
> > You want to add a mailcatcher for tests ? we already have maildev that starts and close on tests files.
> 
> this is not for tests, this is for developers who want to start working easily with minimal fuss by doing basically:
> 
> ```
> git clone ...
> docker-compose up
> ```
> 
> and get a working environment where they can test their changes.

We also use docker for nightly tests, so may i suggest adding profiles to docker-compose https://docs.docker.com/compose/profiles/ for dev and QA profiles.
(not on this PR but the next with the mail catcher)

**Comment by tswfi** [comment]
> We also use docker for nightly tests, so may i suggest adding profiles to docker-compose https://docs.docker.com/compose/profiles/ for dev and QA profiles. (not on this PR but the next with the mail catcher)

This one? https://github.com/PrestaShop/PrestaShop/blob/develop/.github/workflows/cron_nightly_tests_reusable.yml

I have not used profiles in docker compose. Can you give me some pointers on what you mean by adding those profiles? 

In my point of view these changes do not really change anything but fix some of the problems (like using old version of npm and creating extra files because sed parameters are wrong)

I do see how my upcoming pr for mailcatcher might have some overlap with QA runs if they use the same docker-compose file and for that the profiles seem like a perfect solution (e.g. no profile defined start mailcatcher and change the shop smtp settings automatically. if QA profile is set don't start mailcatcher and no changes to smtp configuration). And if `maildev` is the preferred solution I'm happy to change mailcatcher to maildev before I open that PR.

**Comment by boubkerbribri** [comment]
> This one? https://github.com/PrestaShop/PrestaShop/blob/develop/.github/workflows/cron_nightly_tests_reusable.yml
Yes

> I have not used profiles in docker compose. Can you give me some pointers on what you mean by adding those profiles?
> 
> In my point of view these changes do not really change anything but fix some of the problems (like using old version of npm and creating extra files because sed parameters are wrong)
> 
> I do see how my upcoming pr for mailcatcher might have some overlap with QA runs if they use the same docker-compose file and for that the profiles seem like a perfect solution (e.g. no profile defined start mailcatcher and change the shop smtp settings automatically. if QA profile is set don't start mailcatcher and no changes to smtp configuration). And if `maildev` is the preferred solution I'm happy to change mailcatcher to maildev before I open that PR.

I can explain how it's working for tests right now : 
We start docker using docker-compose file and we install `mail-utils` in addition to avoid errors on messages when PrestaShop tries to send an email. When a test needs an email check, we add a pre-test to start maildev (using the node [library](https://www.npmjs.com/package/maildev)) and change the smtp config.

If you add a mailcatcher on docker file, I see 2 solutions : 
- Use profiles like you said, 1 QA without mailcatcher and 1 Dev with it (and we need to change some commands on the workflow that you mentionned).
- Add maildev as mailcatcher, and delete pre-tests from our side (to have maildev everywhere)
I'm more fun of the second solution.

Anyway, for either solutions we must work together on it.


PS: No need to make change on this PR as @nesrineabdmouleh tested it and it's working.

**Comment by tswfi** [comment]
Agreed, I was not aware that this same docker-compose is used in QA runs also (even when I did the pr about removing the `npm setup` steps I didn't notice it).

My idea with maildev / mailcatcher is to have a easy to start dev environment and if the same solution works for QA it would be great. I already have a branch based on this branch that adds mailcatcher and changes the shop settings to use it for emails and I'm just now testing maildev as an alternative to mailcatcher.

Does the QA run also check the email contents somehow? I did a quick peek but couldn't see anything that looks like it does? Looks like maildev has some tools for doing tests for the email contents also but that might require bit more work?

**Comment by boubkerbribri** [comment]
> Does the QA run also check the email contents somehow? I did a quick peek but couldn't see anything that looks like it does? 

Auto tests don't check the value of each email template, but we check that emails are sent when forgetting password (and some use cases like that)

**Comment by Progi1984** [comment]
@PrestaShop/ui-tests-maintainers On my side, it's a GO : https://github.com/Progi1984/ga.tests.ui.pr/actions/runs/4630153706

**Review by Progi1984 — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]
Thanks @tswfi

**Review by ga-devfront — APPROVED** [review]
All work fine for me thank you for your work !

**Review by matthieu-rolland — APPROVED** [review]

**Comment by tswfi** [comment]
image

Not sure what is wanted from me at this point?

**Comment by 0x346e3730** [comment]
Don't mind the Dashboard thing, you should check only the labels, the PR is blocked because of my changes request I will review in a moment to unblock it

**Review by boubkerbribri — APPROVED** [review]

**Comment by boubkerbribri** [comment]
@PrestaShop/qa-functional this is easly tested by the ga.test.ui.pr tool as the new version use the docker-compose file now, but if you have something else in mind, don't hesitate to ping us to help you test it

**Review by HanaRebaiQA — APPROVED** [review]
Hello @tswfi 

I have checked your PR. 

I launched this command `docker-compose up` with mysql v8 
The automatic tests are OK

So, it's QA :heavy_check_mark: 

Thank You!

**Comment by jolelievre** [comment]
Thanks @tswfi and @HanaRebaiQA

**Pull request merged**
